packages/sanity/src/structure/components/requestPermissionDialog/RequestPermissionDialog.tsx

```diff
@@ -0,0 +1,178 @@
+import {Box, Card, DialogProvider, Flex, Stack, Text, TextInput, useToast} from '@sanity/ui'
+import {useEffect, useId, useMemo, useState} from 'react'
+import {type Role, useClient, useProjectId, useTranslation, useZIndex} from 'sanity'
+import {styled} from 'styled-components'
+
+import {type AccessRequest} from '../../../core/studio/screens'
+import {Dialog} from '../../../ui-components'
+import {structureLocaleNamespace} from '../../i18n'
+
+const MAX_NOTE_LENGTH = 150
+
+/** @internal */
+export const DialogBody = styled(Box)`
+  box-sizing: border-box;
+`
+
+/** @internal */
+export const LoadingContainer = styled(Flex).attrs({
+  align: 'center',
+  direction: 'column',
+  justify: 'center',
+})`
+  height: 110px;
+`
+
+/** @internal */
+export interface RequestPermissionDialogProps {
+  onClose?: () => void
+  onRequestSubmitted?: () => void
+}
+
+/**
+ * A confirmation dialog used to prevent unwanted document deletes. Loads all
+ * the referencing internal and cross-data references prior to showing the
+ * delete button.
+ *
+ * @internal
+ */
+export function RequestPermissionDialog({
+  onClose,
+  onRequestSubmitted,
+}: RequestPermissionDialogProps) {
+  const {t} = useTranslation(structureLocaleNamespace)
+  const dialogId = `request-permissions-${useId()}`
+  const projectId = useProjectId()
+  const client = useClient({apiVersion: '2024-09-26'})
+  const toast = useToast()
+  const zOffset = useZIndex()
+
+  const [isSubmitting, setIsSubmitting] = useState(false)
+
+  const [roles, setRoles] = useState<Role[]>()
+
+  const [note, setNote] = useState('')
+  const [noteLength, setNoteLength] = useState<number>(0)
+
+  const [msgError, setMsgError] = useState<string | undefined>()
+  const [hasTooManyRequests, setHasTooManyRequests] = useState<boolean>(false)
+  const [hasBeenDenied, setHasBeenDenied] = useState<boolean>(false)
+
+  //   Get the available roles for the project
+  useEffect(() => {
+    if (!projectId || !client) return
+    client
+      .request({
+        url: `/projects/${projectId}/roles`,
+        method: 'get',
+      })
+      .then((data) => setRoles(data))
+  }, [projectId, client])
+
+  //   Set a default requestedRole based on the available roles
+  const requestedRole = useMemo(() => {
+    const hasEditor = roles?.find((role) => role.name === 'editor')
+    return hasEditor ? 'Editor' : 'Administrator'
+  }, [roles])
```

Comment:
Suggestion only, not necessary, but it could be nice to update this to use rxjs as most of the pieces in the studio.
```
  const requestedRole$: Observable<'Administrator' | 'Editor'> = useMemo(() => {
    const adminRole = 'Administrator' as const
    if (!projectId || !client) return of(adminRole)
    return client.observable.request<Role[]>({url: `/projects/${projectId}/roles`}).pipe(
      map((roles) => {
        const hasEditor = roles.find((role) => role.name === 'editor')
        return hasEditor ? 'Editor' : adminRole
      }),
      startWith(adminRole),
      catchError(() => of(adminRole)),
    )
  }, [projectId, client])

  const requestedRole = useObservable(requestedRole$)

```

if not, we need to add try catch to the request and make the client request promise cancellable in the useEffect

Reply:
Added!